### ISSUE TITLE
Property annotations using `type[]` or `Type[]` and stdlib classes generate an error

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2305]
**To Reproduce**
```python
from unittest import mock


class Thing:
    ...


class ThingProducer:
    def __init__(self) -> None:
        self.produce: type[Thing] = Thing  # no error here


class MockProducer:
    def __init__(self) -> None:
        self.produce: type[mock.Mock] = mock.Mock  # error: Type "Type[Mock]" cannot be assigned to type "Never"
```

**Expected behavior**
No error is generated when typechecking the above code

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line tool

Which version?
1.1.168

**Comment by erictraut** [comment]
Thanks for reporting this. The problem here is specific to `Mock`. It's defined in the `mock.pyi` stub as inheriting from `NonCallableMock`, which in turn inherits from `Any`!?!? I'm not sure why it is defined this way. This effectively blinds the type checker from performing many operations on objects of this class because the class hierarchy is unknown.

I've updated the type checker logic to handle this case in a more elegant manner — one that won't result in a false positive error.

**Comment by JelleZijlstra** [comment]
The inheritance from Any is because mock objects as used in unit tests are often intended to be passed to functions that aren't actually typed to accept them. This is obviously hard to represent in the type system, and the inheritance from Any is what we're using for now.

**Comment by erictraut** [comment]
This is included in pyright 1.1.169. It will also be included in the next release of pylance.

**Comment by bryanforbes** [comment]
There still seems to be issues with these properties. The error is gone from the property declaration, but when I go to use the function:

```python
from unittest import mock

class Thing:
    def __init__(self, arg: str) -> None:
        self.arg: str = arg


class ThingProducer:
    def __init__(self) -> None:
        self.produce: type[Thing] = Thing  # no error here


class MockProducer:
    def __init__(self) -> None:
        self.produce: type[mock.Mock] = mock.Mock


tp = ThingProducer()
tp.produce('asdf')  # no error here

mp = MockProducer()
mp.produce()  # error: Type of "produce" is unknown
```

**Comment by erictraut** [comment]
I could argue this behavior is correct, but I can see why you would expect it to operate otherwise.

The problem is that the `Mock` class derives from `Any`, so the metaclass for `Mock` cannot be established. If the metaclass were to implement a descriptor protocol, then `mp.produce` will get the result of the `__get__` method of the metaclass. Since pyright can't determine the metaclass, it is returning `Unknown`. However, I think it's reasonable to assume that if the metaclass cannot be established, then it's safe to assume that it's `type` or some other class that doesn't implement a descriptor protocol.

I've changed the logic accordingly.
